// txp-utils/lib/node/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLatestLogFilePath = exports.fatal = exports.debug = exports.event = exports.info = exports.ready = exports.warn = exports.error = exports.wait = void 0;
const path_1 = require("path");
const pino_1 = __importDefault(require("pino"));
const chalk_1 = __importDefault(require("chalk"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const loggerDir = (0, path_1.join)(process.cwd(), 'node_modules/.cache/logger');
const loggerPath = (0, path_1.join)(loggerDir, 'txp.log');
const customLevels = {
    ready: 31,
    event: 32,
    wait: 55,
};
let logger;
function init() {
    // 不存在目录就创建它
    if (!logger) {
        fs_extra_1.default.mkdirpSync(loggerDir);
        logger = (0, pino_1.default)({
            customLevels,
            level: 'debug',
        }, pino_1.default.transport({
            targets: [
                {
                    target: 'pino-pretty',
                    options: {
                        destination: loggerPath,
                        colorize: false,
                        translateTime: 'SYS:yyyy-mm-dd HH:MM:ss.l',
                        messageFormat: 'level: {level}、msg: {msg}',
                    },
                    level: 'trace',
                },
            ],
        }));
    }
}
const prefixesBase = {
    wait: chalk_1.default.cyan('wait') + '  -',
    error: chalk_1.default.red('error') + ' -',
    fatal: chalk_1.default.red('fatal') + ' -',
    warn: chalk_1.default.yellow('warn') + '  -',
    ready: chalk_1.default.green('ready') + ' -',
    info: chalk_1.default.cyan('info') + '  -',
    event: chalk_1.default.magenta('event') + ' -',
    debug: chalk_1.default.gray('debug') + ' -',
};
const prefixes = new Proxy(prefixesBase, {
    get(target, key, receiver) {
        init();
        return Reflect.get(target, key, receiver);
    },
});
function wait(...message) {
    console.log(prefixes.wait, ...message);
    logger.wait(message[0]);
}
exports.wait = wait;
function error(...message) {
    console.error(prefixes.error, ...message);
    logger.error(message[0]);
}
exports.error = error;
function warn(...message) {
    console.warn(prefixes.warn, ...message);
    logger.warn(message[0]);
}
exports.warn = warn;
function ready(...message) {
    console.log(prefixes.ready, ...message);
    logger.ready(message[0]);
}
exports.ready = ready;
function info(...message) {
    console.log(prefixes.info, ...message);
    logger.info(message[0]);
}
exports.info = info;
function event(...message) {
    console.log(prefixes.event, ...message);
    logger.event(message[0]);
}
exports.event = event;
function debug(...message) {
    if (process.env.DEBUG) {
        console.log(prefixes.debug, ...message);
    }
    logger.debug(message[0]);
}
exports.debug = debug;
function fatal(...message) {
    console.error(prefixes.fatal, ...message);
    logger.fatal(message[0]);
}
exports.fatal = fatal;
function getLatestLogFilePath() {
    return loggerPath;
}
exports.getLatestLogFilePath = getLatestLogFilePath;
